server/channels/jobs/batch_worker.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package jobs
+
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+)
+
+type BatchWorker[T interface{}] struct {
```

Comment:
`BatchWorker` provides an interface for a job to create a worker that will loop a method over and over at a specified interval, with parameters that the method can modify to execute over a large dataset. The method can also tell the stop to stop when it's done, or the JobService can pause the job and resume with the last saved parameters.

This was refactored out of the work @lieut-data did on the `BatchMigrationWorker`, which now extends this module.

Reply:
At first blush, it's not obvious to me what the generic offers here. It seems like `BatchWorker` itself doesn't use the `app T`, so would it make more sense for any embedding types to just accept an interface useful to itself?

See also https://go.dev/blog/when-generics and the section on `Don’t replace interface types with type parameters`.

Reply:
Yeah this is something I can iterate on a bit. One thing for the `batch_report_worker` is that it does need app methods, and I noticed our pattern was to define the interface and have the job server fill it with what's usually `AppIFace`. Is there a reason we do that? I could just make it require `AppIFace` I suppose.

Reply:
I've removed the generics from BatchWorker, but I can't seem to get around using them for the report worker if I want to be able to define an extended interface on the job itself.

Is there something else I can do with this?